### ISSUE TITLE
HTTP response headers are cached

### ISSUE THREAD

**Issue by xchs** [contao/contao#460]
**Affected version(s)**

Contao 4.4.38
Contao 4.7.4

**Description**

HTTP requests that get an `Content-Disposition: attachment` HTTP response from the web server (indicating the file resource should be downloaded) unexpectedly generate a cache entry with the HTTP response headers, even if the server HTTP cache `var/cache/prod/http_cache` is completely disabled.

Contao 4.7.4 also creates additional cache files.

BTW: I wonder if it is correct that Contao 4.7.4 now saves the cache files to `var/cache/prod/http_cache/http_cache` instead of `var/cache/prod/http_cache` as it has been so far.

**Background story**

Recently, I had to struggle with a client's shared hosting environment which promises unlimited [sic!] web space, but at the same time has a limit for the maximum available inodes (262,144). If you run some Contao 4 installations on this webspace and have activated the HTTP cache for all of them, you will quickly reach the inode limit due to the large number of files (and folders) stored in the cache directory. Therefore, I had to deactivate the HTTP cache completely for some of these installations, because otherwise it was no longer possible to create new files and folders. After switching off the HTTP cache, however, I noticed that some cache entries are still generated for certain requests, which fill up the cache again over time.

**How to reproduce**

- Log into the Contao 4.4 back end of the Contao Official Demo (COD) installation
- (facultative) _Layout_ > _Site structure_: Edit the settings of the website root page and make sure that the "_Server cache timeout_" is set to "_0 (do not cache)_" (which should already be the case in the COD by default)
- (facultative) _System_ > _Maintenance_: Run the _Purge the page cache_ job so that the `var/cache/prod/http_cache` directory should be empty anyway
- Switch to the COD front end and open e.g. the page _Contao Official Demo > Content Elements > File elements_
- Click on the available download resources (most web browsers should present a "_Save as_" dialog)
- What is unexpected at this point: Every (first) request of these resources creates new cache entries even though the server caching should be disabled already

**Comment by fritzmg** [comment]
I can confirm for Contao 4.4.38, that the first request of a session to a file download creates a `http_cache` entry (with no content).

I can also confirm for Contao 4.7.4, that the HTTP cache files are created in `var/cache/prod/http_cache/http_cache` instead of `var/cache/prod/http_cache`. Though you should create a separate issue for that I think.

**Comment by xchs** [comment]
> Though you should create a separate issue for that I think.

I'll do that. But I was unsure whether this behavior was intended or not. Also requests from the Contao back end (`tl_page`, `tl_article`, `tl_content` etc.) now seem to create a cache entry in `var/cache/prod/http_cache/http_cache`

**Comment by Toflar** [comment]
The nested `http_cache` folder is not something you should be worried about. That's just a little blemish that doesn't affect the outcome.

**Comment by Toflar** [comment]
The back end, however, should never create any cache entry (and I cannot reproduce that).

**Comment by xchs** [comment]
> That's just a little blemish that doesn't affect the outcome.

No, it doesn't affect the caching. But should I still file a new ticket for it or not?

**Comment by Toflar** [comment]
Why would you? Doesn't matter how the folder's named. It's not related to Contao anyway :)

**Comment by xchs** [comment]
> The back end, however, should never create any cache entry (and I cannot reproduce that).

To reproduce:

* Log into the Contao back end (Contao 4.7.4)
* _System > Maintenance_: Run the _Purge the page cache_ job so that the `var/cache/prod/http_cache/http_cache` directory is empty (or clear the cache folder with `rm -rf …`)
* _Layout > Site structure_: Create a _New page_
* Check the cache directory `var/cache/prod/http_cache/http_cache`
* `var/cache/prod/http_cache/http_cache/G/A/vSWGZyvAhWjeSvOPrDoA`

```
0
contao.db.tl_page%00tags%00
i:2;
```
  * `var/cache/prod/http_cache/http_cache/X/F/H7vflgu75cw34PF4v1EA`
  ```
0
contao.db.tl_page.4%00tags%00
i:2;
  ```
* Similar cache entries also exist when creating new articles or content elements.

**Comment by Toflar** [comment]
These are all cache tag entries, not real cache entries. I'll have to investigate this at some point.

**Comment by Toflar** [comment]
So what's happening here is that when you save an entry in the back end, we invalidate the cache entries that are associated with it. That's why you get the tags `contao.db.tl_page` and `contao.db.tl_page.4` (because you edited ID 4). Now, the `TagAwareAdapter` of Symfony which `toflar/psr6-symfony-http-cache-store` is using, is a general abstraction not knowing about the cache storage back end (redis, filesystem etc.). So it does not know how to invalidate those two cache entries directly. What it does instead is store these as tag cache entries so when the associated cache entries are **retrieved** they will get invalidated. That's why you get two "cache entries".
Now, what can we do about it? Well, all the fixes are on their way in fact, you just need some patience 😄 

* We need a better, much more efficient `TagAwareAdapter` (a native one) that knows how to invalidate directly. Already on it's way with the upcoming Symfony 4.3: https://github.com/symfony/symfony/pull/30370
* The `toflar/psr6-symfony-http-cache-store` needs to use that as soon as it's released to which I've already created an issue 5 days ago: https://github.com/Toflar/psr6-symfony-http-cache-store/issues/10
* It's nonsense to show the number of files in the back end maintenance screen anyway because it represents not what it is. We'll drop that anyway because we want to support clearing the reverse proxy cache of any cache. PR already here and waiting for 4.8 and dependent releases: https://github.com/contao/contao/pull/173

**Comment by leofeyer** [comment]
> The nested `http_cache` folder is not something you should be worried about. That's just a little blemish that doesn't affect the outcome.

That's correct. I still wonder why this is though.

**Comment by Toflar** [comment]
Because we pass `var/cache/%kernel.environment%/http_cache` to my PSR-6 cache and this one defines another `http_cache` folder by default itself.

**Comment by leofeyer** [comment]
I see. We should still fix this IMHO; probably in the `ContaoKernel` class (?).

**Comment by Toflar** [comment]
I don't care at all to be honest. I won't provide any default path anymore when I release v2 of the `Psr6Store` once Symfony 4.3 has been released.

**Comment by Toflar** [comment]
@xchs and @fritzmg: If you want to test, you can now:

Update your `composer.json` as follows:

```json
"symfony/cache": "dev-master as 4.2.0",
"toflar/psr6-symfony-http-cache-store": "dev-develop as 1.2.0"
```

Note: This is **not** for production! The versioning is wrong and only there to fulfil the rest of the constraints.
When you then add a new page in the back end, nothing should happen. You should not get any tag related cache entries. However, if you do have caching enabled in the FE, it should then create loads of tags in the `var/cache/prod/http_cache/tags` (note: the duplicate `http_cache`'s gone) folder with symlinks and stuff. Invalidation should work just the same as before except for serving should actually also be a lot faster because of a huge decrease in cache lookups on read 😄

**Comment by xchs** [comment]
> When you then add a new page in the back end, nothing should happen. You should not get any tag related cache entries.

Yeah, that works. Thanks 👍

**Comment by xchs** [comment]
> if you do have caching enabled in the FE, it should then create loads of tags in the var/cache/prod/http_cache/tags

This, however, does not seem to work. I have enabled the FE cache (set to 1 year), but I'm getting neither this `tags` subdirectory, nor are the front end pages cached.

```
Package operations: 0 installs, 3 updates, 0 removals
  - Removing symfony/cache (v4.2.7)
  - Installing symfony/cache (dev-master d8dda03): Cloning d8dda03e4a from cache
  - Removing toflar/psr6-symfony-http-cache-store (1.1.2)
  - Installing toflar/psr6-symfony-http-cache-store (dev-develop f4333d8): Cloning f4333d8fcc from cache
  - Updating friendsofsymfony/http-cache (2.5.4 => 2.6.0): Loading from cache
```

**Comment by Toflar** [comment]
Well only if the response is cacheable and there are many reasons why it‘s not. Such as cookies on the request ;) No changes here.

**Comment by xchs** [comment]
> Such as cookies

Yeah, thanks. That was actually the reason. Now it works. 👍

**Comment by leofeyer** [comment]
Since this issue will be resolved once the new versions of `symfony/cache` and `toflar/psr6-symfony-http-cache-store` have been released, I am closing it now.

**Comment by xchs** [comment]
I'm not sure, but the initial problem is still an issue.

**Comment by Toflar** [comment]
What's the initial issue?

**Comment by xchs** [comment]
This one:

> HTTP requests that get an Content-Disposition: attachment HTTP response from the web server (indicating the file resource should be downloaded) unexpectedly generate a cache entry with the HTTP response headers, even if the server HTTP cache var/cache/prod/http_cache is completely disabled.

**Comment by Toflar** [comment]
I couldn't reproduce it. Are you sure the entry is an actual cache entry and not a tag?

**Comment by xchs** [comment]
I'm gonna check it again.

**Comment by xchs** [comment]
This is what happens now (with HTTP cache disabled via website root):
* Loading a front end page --> no caching (`var/cache/prod/http_cache` remains completely empty)
* Initiate a file download --> several tags are created (also the HTTP response headers are saved):

`var/cache/prod/http_cache/2/Y/ebNiqWowbBNfk8jSoyYg`:
```
0
ene3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
a:2:{s:5:"value";b:0;s:4:"tags";a:0:{}}
```

`var/cache/prod/http_cache/R/O/CEeN5ycWEdpVAJ2NCBcQ`:
```
0
write-operations-counter
a:2:{s:5:"value";i:1;s:4:"tags";a:0:{}}
```

`var/cache/prod/http_cache/tags/N/N/WBLEJl9P--ncjIPMkV9Q/V/N/iEKzYjkRPFQN2hgjOY+w` symlinks to `var/cache/prod/http_cache/V/N/iEKzYjkRPFQN2hgjOY+w`

`var/cache/prod/http_cache/tags/T/I/IItsWXvvuM2wwaSfkJow/V/N/iEKzYjkRPFQN2hgjOY+w` symlinks to `var/cache/prod/http_cache/V/N/iEKzYjkRPFQN2hgjOY+w`

`var/cache/prod/http_cache/tags/T/Z/RYaNcq85p9IvE2idVe9w/V/N/iEKzYjkRPFQN2hgjOY+w` symlinks to `var/cache/prod/http_cache/V/N/iEKzYjkRPFQN2hgjOY+w`

`var/cache/prod/http_cache/V/N/iEKzYjkRPFQN2hgjOY+w`:
```
0
md13a4a4439af3306a055a1d3e0ed9d96dc7bd18cd6a32ff226f42040ff4c52956
a:2:{s:5:"value";a:1:{s:11:"non-varying";a:3:{s:4:"vary";a:0:{}s:7:"headers";a:15:{s:13:"cache-control";a:1:{i:0;s:50:"must-revalidate, post-check=0, pre-check=0, public";}s:4:"date";a:1:{i:0;s:29:"Tue, 30 Apr 2019 14:45:24 GMT";}s:13:"last-modified";a:1:{i:0;s:29:"Tue, 02 Jan 2018 10:37:38 GMT";}s:19:"content-disposition";a:1:{i:0;s:37:"attachment; filename=foobar-12345.pdf";}s:10:"connection";a:1:{i:0;s:5:"close";}s:12:"x-powered-by";a:1:{i:0;s:9:"PHP/7.3.2";}s:22:"x-content-type-options";a:1:{i:0;s:7:"nosniff";}s:15:"referrer-policy";a:1:{i:0;s:59:"no-referrer-when-downgrade, strict-origin-when-cross-origin";}s:12:"content-type";a:1:{i:0;s:15:"application/pdf";}s:14:"content-length";a:1:{i:0;i:0;}s:13:"accept-ranges";a:1:{i:0;s:5:"bytes";}s:15:"x-frame-options";a:1:{i:0;s:10:"SAMEORIGIN";}s:16:"x-xss-protection";a:1:{i:0;s:13:"1; mode=block";}s:12:"x-cache-tags";a:1:{i:0;s:68:"contao.db.tl_article.1,contao.db.tl_content.1,contao.db.tl_content.2";}s:16:"x-content-digest";a:1:{i:0;s:66:"ene3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855";}}s:6:"status";i:200;}}s:4:"tags";a:3:{s:22:"contao.db.tl_article.1";s:22:"contao.db.tl_article.1";s:22:"contao.db.tl_content.1";s:22:"contao.db.tl_content.1";s:22:"contao.db.tl_content.2";s:22:"contao.db.tl_content.2";}}
```

So this single file download now requires a total of 27 inodes.

**Comment by Toflar** [comment]
The inodes are just fine as explained before. However, the file should not be cached. Damn, the `BinaryFileResponse` makes downloads `public` by default. What the heck.

Can you try chainging the code here: https://github.com/contao/contao/blob/master/core-bundle/src/Resources/contao/library/Contao/File.php#L772

Do this:

```diff
$response = new BinaryFileResponse($this->strRootDir . '/' . $this->strFile);
+ $response->setPrivate(); // BinaryFileResponse is public by default
+ $response->setAutoEtag();

$response->setContentDisposition
(
    $inline ? ResponseHeaderBag::DISPOSITION_INLINE : ResponseHeaderBag::DISPOSITION_ATTACHMENT,
    $filename,
    Utf8::toAscii($this->basename)
);

$response->headers->addCacheControlDirective('must-revalidate');
- $response->headers->addCacheControlDirective('post-check', 0);
- $response->headers->addCacheControlDirective('pre-check', 0);
- $response->headers->set('Connection', 'close');

throw new ResponseException($response);
```

**Comment by xchs** [comment]
Thanks for tracking it down. That did the trick. No cache entries anymore.

**Comment by Toflar** [comment]
@leofeyer, this needs to be reopened and assigned to 4.4 then :)

**Comment by Toflar** [comment]
BTW: Here's the reasoning behind what I did

* `setPrivate()` - well that one is the bug
* `setAutoEtag()` - generates an ETag so that when you have already downloaded the exact same file, it's not transferred anymore (`304 Not Modified`)
* `post-check` and `pre-check` are some IE left-overs. Go search the internet, I couldn't find what the purpose of those were
* `Connection: close` is not needed imho. It makes no sense to me, sounds like an old workaround for some old browser too.

**Comment by ausi** [comment]
But adding `setPrivate()` would also bypass the cache for sites with an active cache, wouldn’t it?

**Comment by Toflar** [comment]
Which is good imho. I don‘t see why - in our case - we would want to store the files in the reverse proxy.

**Comment by leofeyer** [comment]
I think we should keep the `Connection: close` header so we do not run into problems like this: https://github.com/guzzle/guzzle/issues/1977

**Comment by leofeyer** [comment]
Fixed in 04112467be89fce78b375bebe329d80e3e968f2e. Thank you @Toflar.
